aten/src/ATen/native/Pad.cpp

```diff
@@ -0,0 +1,271 @@
+              at::indexing::None)));
+        }
+      }
+      result.index_put_(slices, constant_values[pad_dim_idx][1]);
+    }
+  }
+}
+
+void check_pad_specifier_is_none(const c10::optional<Tensor>& arg, const char* arg_name, c10::string_view mode_str) {
+  TORCH_CHECK(!arg.has_value(),
+    "Unsupported keyword argument for '", std::string(mode_str), "' mode: ", arg_name);
+}
+
+// Calculate the size of the padded tensor
+//
+// NOTE: pad_width must be processed with _expand_pad_specifier before
+// calling this function
+Tensor get_result_size(const Tensor& self, const Tensor& pad_width) {
+  auto result_size_tensor = at::tensor(
+    self.sizes(),
+    TensorOptions().dtype(at::kLong).device(at::kCPU).memory_format(c10::MemoryFormat::Contiguous));
+
+  result_size_tensor += pad_width.sum(1).cpu();
+
+  return result_size_tensor;
+}
+
+// Create an IntArrayRef from a tensor
+IntArrayRef tensor_to_arrayref(const Tensor& size_tensor) {
+  return IntArrayRef(size_tensor.data_ptr<int64_t>(), size_tensor.numel());
+}
+
+// Check that a tensor is on the expected device
+void check_device(const Tensor& arg, const char* arg_name, at::Device self_device) {
+  TORCH_CHECK(arg.device() == self_device,
+    "Expected '", arg_name, "' to be on the same device as 'self' (",
+    self_device, ") but got ", arg.device());
+}
+
+Tensor& pad_out_impl(
+  const Tensor& self,
+  const Tensor& pad_width,
+  c10::string_view mode_str,
+  const c10::optional<Tensor>& constant_values_opt,
+  Tensor& result
+) {
+  // pad_width must be Long and on CPU
+  TORCH_CHECK(pad_width.device() == at::kCPU,
+    "Expected 'pad_width' to be on CPU, but got ", pad_width.device());
+  TORCH_CHECK(pad_width.scalar_type() == at::ScalarType::Long,
+    "Expected 'pad_width' to be Long dtype, but got ",
+    pad_width.scalar_type());
+
+  PadMode mode = get_pad_mode_from_str(mode_str);
+
+  if (mode != PadMode::Constant) {
+    // constant_values should be none if not using constant mode
+    check_pad_specifier_is_none(constant_values_opt, "constant_values", mode_str);
+  } else {
+    // If using constant_values is given, it should be on same device as self
+    if (constant_values_opt.has_value()) {
+      check_device(constant_values_opt.value(), "constant_values", self.device());
+    }
+  }
+
+  // If `out` is given, it must match `self`'s dtype and device
+  if (result.defined()) {
+    TORCH_CHECK(result.scalar_type() == self.scalar_type(),
+      "Expected 'out' dtype ", self.scalar_type(), " but got ", result.scalar_type());
+    check_device(result, "out", self.device());
+  }
+
+  Tensor pad_width_ = at::native::_expand_pad_specifier(pad_width, "pad_width", self.dim());
+
+  auto result_size_tensor = get_result_size(self, pad_width_);
+  auto result_size = tensor_to_arrayref(result_size_tensor);
+  if (result.defined()) {
+    at::native::resize_output(result, result_size);
+  } else {
+    result = at::empty(result_size, self.options());
```

Comment:
Potential optimization if constant values is just a single scalar is to use full here? Maybe a future TODO?

Reply:
Yeah that should make the scalar case faster. I'll add a comment here